### PR TITLE
linter: added checks for functions

### LINKED ISSUE #561: Find mismatching @return annotations
## Code Example

```php
    /**
     * Set the current local namespace.
     * @param array|null $namespace (default: null)
     * @return array|null
     */
    public function setNamespace(array $namespace = null)
    {
        $this->namespace = $namespace;
    }
```

## Actual Behavior

No warnings.

## Expected Behavior

Warning about not returning anything from a function that promised to return something (except null) in its phpdoc.

### LINKED ISSUE #563: For non-void functions, forbid bare return
## Code Example

```php
/**
 * @return string|null
 */
function f($x) {
  if (!$x) {
    return; // Better to write as "return null"
  }
  // ...
}
```

## Actual Behavior

No warnings.

## Expected Behavior

Suggest explicit `return null`. Functions with void results should be ignored.

### LINKED ISSUE #597: Store FQN in CurrentFunction?
Right now `ClassParseState.CurrentFunction` stores only a partial (base) function name instead of FQN.

It's probably more useful to store FQN there.

Since we store method names there as well, it might be a good time to move that to the `CurrentMethod` field? We can't store method FQN in a way that is useful for us.

This needs some research on how CurrentFunction is used. It should be trivial to get the base name out of the FQN when it's needed.

### PR DESCRIPTION
**Pull request by i582**
1. Added check for bare returns for functions with non-void return types. #563
2. Added check for the absence of return statement in function with `@return` annotation, which is not equal to void. #561

For this, the storage of `CurrentFunction` has been improved.
`CurrentFunction` now stores the fully qualified name of the function with its class and namespace. A separate `CurrentMethod` field was created to store the names of the methods. #597 

Fixes #561 
Fixes #563 
Fixes #597

### CODE REVIEW

File: src/linter/block.go [hunk 1]
```diff
@@ -64,10 +64,14 @@ type BlockWalker struct {
 	// state
 	statements map[node.Node]struct{}
 
-	// whether a function has a return without explit expression.
+	// whether a function has a return without explicit expression.
 	// Required to make a decision in void vs null type selection,
 	// since "return" is the same as "return null".
 	bareReturn bool
+	// when the function contains at least one return or yield.
+	containsReturnOrYield bool
```
Comment (YuriyNasretdinov) [inline src/linter/block.go:72]: exitFlags should already contain this information in the way you want?

File: src/linter/block.go [hunk 2]
```diff
@@ -525,9 +544,50 @@ func (b *BlockWalker) handleFunction(fun *stmt.Function) bool {
...
+			if !ok {
+				return
+			}
+
+			if !fun.Typ.Is("void") {
+				b.r.Report(ret, LevelWarning, "bareReturn", "Replace 'return' with 'return null'")
+			}
+		}
+
+		// for _, el := range meta.Info.allFunctions.H {
```
Comment (YuriyNasretdinov) [inline src/linter/block.go:587]: Probably remove this?

File: src/linter/root.go [hunk 3]
```diff
@@ -698,7 +698,12 @@ func (d *RootWalker) handleFuncStmts(params []meta.FuncParam, uses, stmts []node
 		prematureExitFlags = cleanFlags
 	}
 
+	// if the function contains only one statement and this statement is throw.
+	oneThrowStatementFunction := b.containsThrow && len(stmts) == 1
```
Comment (YuriyNasretdinov) [inline src/linter/root.go:702]: Why not two? Ten? Why this special case only?

File: src/linttest/testdata/mustache/golden.txt [hunk 4]
```diff
@@ -52,6 +52,21 @@ MAYBE   phpdoc: Missing PHPDoc for "getHelperName" public method at testdata/mus
 MAYBE   phpdoc: Missing PHPDoc for "getTemplateName" public method at testdata/mustache/src/Mustache/Exception/UnknownTemplateException.php:34
     public function getTemplateName()
                     ^^^^^^^^^^^^^^^
+WARNING bareReturn: Replace 'return' with 'return null' at testdata/mustache/src/Mustache/Parser.php:207
```
Comment (YuriyNasretdinov) [inline src/linttest/testdata/mustache/golden.txt:55]: return; and return null; already mean the same thing. Maybe provide more context here on why exactly do you ask to replace it in this function?

File: src/linter/root.go (hunk 3 above)
Comment (i582) [reply src/linter/root.go:702]: in `basic_test.go`:

```
// Functions of this kind are very common in Symfony, so it makes sense not to consider this 
// a loss of return, although in fact the @return annotation is not entirely correct.
```

File: src/linter/block.go (hunk 1 above)
Comment (i582) [reply src/linter/block.go:72]: I need to know if there is a `return` or `yield` in the body of the function in order to accurately determine `"mismatchingReturn"`.  I'm not sure that exitFlags holds the information I need. Comment mentions other things `// if function has exit / die / throw, then ExitFlags will be <> 0`.

File: src/linttest/testdata/mustache/golden.txt (hunk 4 above)
Comment (i582) [reply src/linttest/testdata/mustache/golden.txt:55]: Probably we can provide additional information after the first sentence?
For example:
```
<critical> WARNING bareReturn: Replace 'return' with 'return null'. The '\f' function has @return annotation with non-'void' value, so 'return null' is preferred. at C:/projects/noverify/example/index.php:7
    return;
    ^^^^^^^
```

**Comment by i582** [comment]
Closed until better times.

**Pull request closed without merging**